pkg/cloudprovider/vsphere/nodemanager.go

```diff
@@ -260,81 +256,94 @@ func (nm *NodeManager) DiscoverNode(nodeID string, searchBy cm.FindVM) error {
-						klog.V(2).Infof("Adding Internal IP by AddressMatching: %s", ip)
-						v1helper.AddToNodeAddresses(&addrs,
-							v1.NodeAddress{
-								Type:    v1.NodeInternalIP,
-								Address: ip,
-							},
-						)
-					}
-
-					if externalNetworkSubnet != nil && externalNetworkSubnet.Contains(parsedIP) {
-						klog.V(2).Infof("Adding External IP by AddressMatching: %s", ip)
-						v1helper.AddToNodeAddresses(&addrs,
-							v1.NodeAddress{
-								Type:    v1.NodeExternalIP,
-								Address: ip,
-							},
-						)
-					}
+			for _, ip := range ips {
+				parsedIP := net.ParseIP(ip)
+				if parsedIP == nil {
+					return fmt.Errorf("can't parse IP: %s", ip)
 				}
-			} else if internalVMNetworkName != "" && strings.EqualFold(internalVMNetworkName, v.Network) {
-				for _, ip := range ips {
-					klog.V(2).Infof("Adding Internal IP by NetworkName: %s", ip)
+
+				// prioritize address masking over networkname
+				if !foundInternal && internalNetworkSubnet != nil && internalNetworkSubnet.Contains(parsedIP) {
+					klog.V(2).Infof("Adding Internal IP by AddressMatching: %s", ip)
 					v1helper.AddToNodeAddresses(&addrs,
 						v1.NodeAddress{
 							Type:    v1.NodeInternalIP,
 							Address: ip,
 						},
 					)
-					found = true
-					break
+					foundInternal = true
 				}
-			} else if externalVMNetworkName != "" && strings.EqualFold(externalVMNetworkName, v.Network) {
-				for _, ip := range ips {
-					klog.V(2).Infof("Adding External IP by NetworkName: %s", ip)
+				if !foundExternal && externalNetworkSubnet != nil && externalNetworkSubnet.Contains(parsedIP) {
+					klog.V(2).Infof("Adding External IP by AddressMatching: %s", ip)
 					v1helper.AddToNodeAddresses(&addrs,
 						v1.NodeAddress{
 							Type:    v1.NodeExternalIP,
 							Address: ip,
 						},
 					)
-					found = true
-					break
+					foundExternal = true
 				}
-			} else {
-				for _, ip := range ips {
-					klog.V(2).Infof("Adding IP: %s", ip)
+
+				// then use network name
+				if !foundInternal && internalVMNetworkName != "" && strings.EqualFold(internalVMNetworkName, v.Network) {
+					klog.V(2).Infof("Adding Internal IP by NetworkName: %s", ip)
 					v1helper.AddToNodeAddresses(&addrs,
 						v1.NodeAddress{
-							Type:    v1.NodeExternalIP,
-							Address: ip,
-						}, v1.NodeAddress{
 							Type:    v1.NodeInternalIP,
 							Address: ip,
 						},
 					)
-					found = true
-					break
+					foundInternal = true
 				}
+				if !foundExternal && externalVMNetworkName != "" && strings.EqualFold(externalVMNetworkName, v.Network) {
+					klog.V(2).Infof("Adding External IP by NetworkName: %s", ip)
+					v1helper.AddToNodeAddresses(&addrs,
+						v1.NodeAddress{
+							Type:    v1.NodeExternalIP,
+							Address: ip,
+						},
+					)
+					foundExternal = true
+				}
+			}
+
+			// At least one of the Internal or External addresses has been found.
+			// Both technically need to exist for the node to function correctly.
```

Comment:
I think it's valid for only one of them to be defined, preferable InternalIP.

Reply:
Will update comment